File: benchmarks/solutions/go/r1-word-freq.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
	"unicode"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "error: no file argument")
		os.Exit(1)
	}

	data, err := os.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	counts := make(map[string]int)
	for _, word := range strings.FieldsFunc(strings.ToLower(string(data)), func(c rune) bool {
		return !unicode.IsLetter(c)
	}) {
		counts[word]++
	}

	type pair struct {
		word  string
		count int
	}
	pairs := make([]pair, 0, len(counts))
	for w, c := range counts {
		pairs = append(pairs, pair{w, c})
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].count != pairs[j].count {
			return pairs[i].count > pairs[j].count
		}
		return pairs[i].word < pairs[j].word
	})

	for i := 0; i < 10 && i < len(pairs); i++ {
		fmt.Printf("%s: %d\n", pairs[i].word, pairs[i].count)
	}
}

File: benchmarks/solutions/haskell/r1-word-freq.hs
module Main where

import Data.Char (isAlpha, toLower)
import Data.List (sortBy)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Ord (Down(..), comparing)
import System.Environment (getArgs)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)

main :: IO ()
main = do
    args <- getArgs
    case args of
        [] -> hPutStrLn stderr "error: no file argument" >> exitFailure
        (path:_) -> do
            text <- readFile path
            let words = filter (not . null) $ splitNonAlpha (map toLower text)
                counts = Map.fromListWith (+) [(w, 1 :: Int) | w <- words]
                sorted = take 10 $ sortBy (comparing (Down . snd) <> comparing fst)
                                          (Map.toList counts)
            mapM_ (\(w, c) -> putStrLn $ w ++ ": " ++ show c) sorted

splitNonAlpha :: String -> [String]
splitNonAlpha [] = []
splitNonAlpha s = let (word, rest) = span isAlpha (dropWhile (not . isAlpha) s)
                 in word : splitNonAlpha rest
